<?xml version="1.0" ?>
<launch>

	<!-- GENERAL ARGUMENTS -->
	<arg name="input_type" default="depth"/>

	<!-- MEASUREMENT SERVER -->
	<param name="measurementServerName" value="/rtabmap_server"/>
    <node pkg="robot_arm_3d_scan" type="RTABMAPServer" name="rtabmap_server" output="screen"/>

	<!-- RTABMAP PROCESSING -->
	<include file="$(find rtabmap_launch)/launch/rtabmap.launch">

		<arg name="rtabmap_viz"		value="true"/>
		<arg name="use_sim_time"    value="true"/>
		<arg name="database_path"	value="$(find robot_arm_3d_scan)/config/rtabmap/rtabmap.db"/>
		<arg name="args" value="--delete_db_on_start"/> <!-- Delete database on start -->

		<!-- Input topics are synchronized -->
		<arg name="approx_sync"				value="false"/>

		<arg name="stereo" 			value="false"/>
		<arg name="depth" 			value="$(eval arg('input_type') == 'depth')"/>

		<param name="rtabmap/map_always_update" type="bool" value="true"/>

		<!-- RTABMAP frames -->
		<arg name="frame_id"					value="camera_link"/> <!--Fixed base frame of the arm-->
		<!--<arg name="odom_frame_id"			value="world"/>--> <!-- Same as frame_id, as the base of the arm doesn t move -->
		
		<!-- RTABMAP camera topics -->
		<arg name="rgb_topic"       		value="/color"/>
		<arg name="depth_topic"     		value="/aligned_depth"/>
		<arg name="camera_info_topic" 		value="/color_camera_info"/>

		<arg name="gen_cloud" value="true" 	if="$(eval arg('input_type') == 'depth')"/>
		<arg name="gen_cloud" value="false" unless="$(eval arg('input_type') == 'depth')"/>

		<arg name="subscribe_scan_cloud"    value="true"/>
		<arg name="scan_cloud_topic"        value="/scan_cloud" if="$(eval arg('input_type') == 'depth')"/>
		<arg name="scan_cloud_topic"        value="/point_cloud" if="$(eval arg('input_type') == 'point_cloud')"/>
		<arg name="gen_cloud_decimation"    value="0"/>
		<arg name="gen_cloud_voxel"         value="0"/>

		<arg name="visual_odometry"         value="true"	if="$(eval arg('input_type') == 'depth')"/>
		<arg name="visual_odometry"         value="false"	unless="$(eval arg('input_type') == 'depth')"/>

		<arg name="icp_odometry"            value="true" 	if="$(eval arg('input_type') == 'point_cloud')"/>
		<arg name="icp_odometry"            value="false"	unless="$(eval arg('input_type') == 'point_cloud')"/>

		<arg name="scan_cloud_assembling"              			value="true"/>
		<arg name="scan_cloud_assembling_fixed_frame"  			value=""/>
		<arg name="scan_cloud_assembling_voxel_size"   			default="0.0"/>
		<arg name="scan_cloud_assembling_range_min"    			default="0.0"/>
		<arg name="scan_cloud_assembling_range_max"    			default="0.0"/>
		<arg name="scan_cloud_assembling_noise_radius"   		default="0.0"/>
		<arg name="scan_cloud_assembling_noise_min_neighbors"   default="0"/>

		<!-- RTABMAP default parameters -->
		<arg name="cfg" value="$(find robot_arm_3d_scan)/config/rtabmap/rtabmap.ini"/>
	</include>

</launch>